common/networkselectionmodel.cpp

```diff
@@ -43,87 +46,174 @@ static void writeSelection(Message *msg, const QItemSelection &selection)
   }
 }
 
+// find a model having a defaultSelectedItem method
+static QAbstractItemModel *findSourceModel(QAbstractItemModel *model)
+{
+  if (model) {
+    if (model->metaObject()->indexOfMethod(QMetaObject::normalizedSignature("defaultSelectedItem()")) != -1) {
+      return model;
+    }
+    else if (model->inherits("QAbstractProxyModel")) {
+      return findSourceModel(qobject_cast<QAbstractProxyModel*>(model)->sourceModel());
+    }
+  }
+
+  return 0;
+}
+
 NetworkSelectionModel::NetworkSelectionModel(const QString &objectName, QAbstractItemModel* model, QObject* parent):
   QItemSelectionModel(model, parent),
   m_objectName(objectName),
   m_myAddress(Protocol::InvalidObjectAddress),
   m_handlingRemoteMessage(false)
 {
+  setObjectName(m_objectName + QLatin1String("Network"));
   connect(this, SIGNAL(currentChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(currentColumnChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentColumnChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(currentRowChanged(QModelIndex,QModelIndex)), this, SLOT(slotCurrentRowChanged(QModelIndex,QModelIndex)));
   connect(this, SIGNAL(selectionChanged(QItemSelection,QItemSelection)), this, SLOT(slotSelectionChanged(QItemSelection,QItemSelection)));
-
-  Q_ASSERT(model);
-  connect(model, SIGNAL(modelAboutToBeReset()), this, SLOT(clearPendingSelection()));
-  connect(model, SIGNAL(rowsInserted(QModelIndex,int,int)), this, SLOT(applyPendingSelection()));
 }
 
 NetworkSelectionModel::~NetworkSelectionModel()
 {
 }
 
-Protocol::ItemSelection GammaRay::NetworkSelectionModel::readSelection(const GammaRay::Message& msg)
+void NetworkSelectionModel::requestSelection()
 {
-    Protocol::ItemSelection selection;
-    qint32 size = 0;
-    msg.payload() >> size;
-    selection.reserve(size);
-
-    for (int i = 0; i < size; ++i) {
-        Protocol::ItemSelectionRange range;
-        msg.payload() >> range.topLeft >> range.bottomRight;
-        selection.push_back(range);
+  if (m_handlingRemoteMessage || !Endpoint::isConnected() || m_myAddress == Protocol::InvalidObjectAddress) {
+    return;
+  }
+  Message msg(m_myAddress, Protocol::SelectionModelStateRequest);
+  Endpoint::send(msg);
+}
+
+void NetworkSelectionModel::sendSelection()
+{
+  if (!Endpoint::isConnected() || m_myAddress == Protocol::InvalidObjectAddress) {
+    return;
+  }
+
+  clearPendingSelection();
+
+  const QItemSelectionModel::SelectionFlags selectionFlags = QItemSelectionModel::ClearAndSelect |
+      QItemSelectionModel::Rows | QItemSelectionModel::Current;
+  const Qt::MatchFlags matchFlags = Qt::MatchExactly | Qt::MatchRecursive;
+
+  if (!hasSelection()) {
+    // Only match tools selection: com.kdab.GammaRay.XXX.selection
+    if (QRegExp(QStringLiteral("com\\.kdab\\.GammaRay\\.[\\w]+\\.selection")).exactMatch(m_objectName)) {
```

Comment:
This looks very fragile, as there is no clear policy on how objects are named. It's for example not matching the entity and frame graph node trees of the Qt3D inspector, which are candidates for default selection.

Isn't the following check for defaultSelectedItem already excluding anything that doesn't want default selection? If not, could we do this in a similar way, e.g. have a dynamic property to indicate no default selection, should this actually be needed somewhere?